Enhance uri utils

RandomShaper opened the pull request:
- Enhance `HTTPClient.query_string_from_dict()`. (See updated code & docs.)
- Unify http- and percent- encode/decode version which was not exposed to scripting (only to GDNative).
This commit keeps the percent-prefixed versions, but with the http-prefixed implementations.

**UPDATE:**

Added:
- Fix buggy percent-encoding. Fixes #17875.

RandomShaper committed:
Enhance HTTPClient.query_string_from_dict()

akien-mga commented:
CC @Faless

akien-mga commented on core/variant_call.cpp:
```diff
@@ -284,8 +284,8 @@ struct _VariantCall {
 	VCALL_LOCALMEM0R(String, json_escape);
-	VCALL_LOCALMEM0R(String, percent_encode);
-	VCALL_LOCALMEM0R(String, percent_decode);
+	VCALL_LOCALMEM0R(String, uri_encode);
```
This breaks compatibility with 3.0, so we can't merge it as is. You can add a `uri_encode`, but the old methods should stay (they can be marked as deprecated though and removed later on).

RandomShaper replied:
If 3.0 and 3.1 have to be compatible, then I think it's better to keep the percent_encode name. A bit less appealing to me, but not bad either.

RandomShaper committed:
Unify http- and percent- encode/decode

There was a percent-prefixed version, which was exposed, and a http-prefixed version which was not (only to GDNative).
This commit keeps the percent-prefixed versions, but with the http-prefixed implementations.

RandomShaper commented:
I've updated this PR. The only incompatibility now is removing `http_(de/en)code` from GDNative.

RandomShaper committed:
Fix buggy percent-encoding

Fixes #17875.

RandomShaper commented:
Please note this is potentially cherry-pickable for 3.0.

VitaZheltyakov commented:
I am test this commits - they not fix https://github.com/godotengine/godot/issues/17875

Now:
```
var dic = { 'arg1' : 'Ж', 'arg2' : foo }
print(http.query_string_from_dict(dic))
```

Return `arg1==%FFDﾨ￵AY%FF9ﾨ￵AY&arg2=foo.` This is wrong.
Need: a`rg1=%D0%96&arg2=foo`

RandomShaper commented:
Under which Godot version?

akien-mga commented:
I'm now getting many `unicode error: invalid skip` errors when running Godot (just the project manager).

A breakpoint on 
```
core/ustring.cpp
1338:                                   _UNICERROR("invalid skip");
```
gives me this backtrace for the first occurrence:
```
(gdb) bt
#0  String::parse_utf8 (this=0x7fffffffbe78, p_utf8=0x6406e90 "/home/akien/Vid0\251os", p_len=-1) at core/ustring.cpp:1338
#1  0x0000000002a608c0 in String::utf8 (p_utf8=0x6406e90 "/home/akien/Vid0\251os", p_len=-1) at core/ustring.cpp:1287
#2  0x0000000002a66c98 in String::percent_decode (this=0x63fb3f0) at core/ustring.cpp:3211
#3  0x00000000016d2ada in _get_drives (list=0x7fffffffbee8) at drivers/unix/dir_access_unix.cpp:247
#4  0x00000000016d2c41 in DirAccessUnix::get_drive_count (this=0x638e160) at drivers/unix/dir_access_unix.cpp:264
#5  0x0000000001fb833c in FileDialog::_update_drives (this=0x636f830) at scene/gui/file_dialog.cpp:742
#6  0x0000000001fb7eea in FileDialog::set_access (this=0x636f830, p_access=FileDialog::ACCESS_FILESYSTEM) at scene/gui/file_dialog.cpp:683
#7  0x0000000001930063 in ProjectManager::ProjectManager (this=0x59f2f30) at editor/project_manager.cpp:1666
#8  0x0000000000f46c1b in Main::start () at main/main.cpp:1686
#9  0x0000000000f169a8 in main (argc=1, argv=0x7fffffffda38) at platform/x11/godot_x11.cpp:54
```
Other occurrences are either the same, or similar but with `EditorFileDialog`.

akien-mga commented:
@VitaZheltyakov Can't reproduce your issue on fa97ec23.
```
	var http = HTTPClient.new()
	var dic = { 'arg1' : 'Ж', 'arg2' : 'foo' }
	print(http.query_string_from_dict(dic))
```
prints:
```
arg1=%D0%96&arg2=foo
```

RandomShaper commented:
What's the actual name of that folder?

RandomShaper commented:
How do file:// entries look in your .config/gtk-3.0/bookmarks?

Noshyaar commented:
@akien-mga Can't reproduce on 48890b1, d87307d no invalid skip errors printed on the project manager. I added `Vidéos` and `Téléchargements` to my bookmark.

akien-mga commented:
> I added Vidéos and Téléchargements to my bookmark.

How are they encoded in your `~/.config/gtk-3.0/bookmarks`?

RandomShaper commented:
I'll have a look very soon.

I can say in advance that the implementation of percent_decode() formerly
used (now replaced by the body of http_decode()) handled this better.

El mié., 11 abr. 2018 16:21, Poommetee Ketson <notifications@github.com>
escribió:

> ~ >>> cat .config/gtk-3.0/bookmarks
> [...]
> file:///home/noshyaar/Desktop/T%C3%A9l%C3%A9chargements
> file:///home/noshyaar/Desktop/Vid%C3%A9os
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/17583#issuecomment-380469669>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCtq5d_b8uuZa-sE2Rm77EnTW45Tjzks5tnhFogaJpZM4Suyvv>
> .
>

VitaZheltyakov commented:
@RandomShaper I am use Godot 2.1.x for Windows.

I have a suspicion that the problem in the function `utf8()`

endragor commented:
This code breaks compatibility for GDNative libraries with no way for binding code to detect it. You shouldn't delete existing methods from gdnative_api.json, and there wasn't really a reason to do it in this case.

endragor commented:
To clarify more - GDNative API supports versioning, and if you absolutely must break something, you are supposed to increase the major version of the API, so newer Godot will not return the older version and binding libraries could be adopted to work with both major versions. But obviously it adds work for each binding library author, so it should be avoided whenever possible.

In this case there was no version change - the methods were deleted from an already released version. That must never be done.

RandomShaper commented:
For 3.0 I will readd the http-prefixed versions.

For 3.1, I'll recheck to confirm http-prefixed and uri-prefixed were doing
the same. In that case, the removal would be interesting to remove
redundant methods.

El jue., 12 abr. 2018 11:59, Ruslan Mustakov <notifications@github.com>
escribió:

> To clarify more - GDNative API supports versioning, and if you absolutely
> must break something, you are supposed to increase the major version of the
> API, so newer Godot will not return the older version and binding libraries
> could be adopted to work with both major versions. But obviously it adds
> work for each binding library author, so it should be avoided whenever
> possible.
>
> In this case there was no version change - the methods were deleted from
> an already released version. That must never be done.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/17583#issuecomment-380747243>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCthz6-M4jZXvI2dZoN2S-KgVv_DBlks5tnyWFgaJpZM4Suyvv>
> .
>

endragor commented:
There is no point to remove them from GDNative API and bump the major API version just because there are redundant methods. As I said, you would be inducing work on other people for no good reason. Keep in mind binding libraries would have to support **BOTH** API versions, so you don't really improve anything by removing redundant stuff, you only add burden.

karroffel commented:
endragor is right, this breaks binary compatibility with libraries, this must not be done like that. A fallback must be provided, we could add a deprecation warning for now, but functions can't just be removed like that, this has to be reverted.

RandomShaper commented:
OK. I wasn't aware of such problems. My apologies.

In a few hours I'm submitting a PR reverting the bad commit to restore the binding and fixing all the problems.

RandomShaper commented:
Reverted at #18156.
Removed the _cherrypick_ tag and submitting instead a variant of this PR for 3.0 excluding the bad commit (#18157).
